### PR TITLE
YIN algorithm, conversion of GuitarTuner example to Swift 3

### PR DESCRIPTION
**Pull request by glaurent**
The GuitarTuner example has also been changed to use the YIN algorithm, which, from the testing I've done (electric and acoustic guitar), gives better results than HPS. The correct note is found on the whole range of the instrument.

### CODE REVIEW

File: Source/Estimation/YINUtil.swift [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+//
+
+import UIKit
+import Accelerate
+
+class YINUtil {
+
+  // slow and eats a lot of CPU, but working
+  //
+  class func difference2(buffer:[Float]) -> [Float] {
```
Comment (vadymmarkov) [inline Source/Estimation/YINUtil.swift:18]: @glaurent Is there any reason of having `class` functions here? Or if it's meant to be a set of functions, could be a struct with static `func`, right?

**Review by vadymmarkov — APPROVED** [review]
@glaurent This is just amazing! I will test it with my guitars and probably make a little cleanup to keep code base more consistent. It also deserves to be in REAME, so I'll update it as well. Thanks for your investment and contribution!

**Comment by PCrompton** [comment]
@glaurent @vadymmarkov Thanks for all your work on this! I've been caught up in other projects right now but I'm excited to get back to my project involving this framework. I'd love to contribute to this project at some point but I need to better learn the concepts behind the code. Do you recommend any books or other resources that will help me understand how to parse sound waves? I have a music degree but aspire to get into the software development field professionally, and my dream is to help develop software for musicians and music education. Thanks again!

**Comment by glaurent** [comment]
@vadymmarkov You're welcome. Thank you for your work on Beethoven, it has helped me immensely too.

As for your question, yes this could be a struct with static func, I'm not sure how much of a difference that would make as far as the language is concerned. I wrote it this way because the original C++ code is written that wayl, and it feels somewhat more natural to me too. I noticed you seem to prefer structs to classes, though :)

BTW, is there a reason to have the TransformStrategy user-settable in the Config class ? As far as I can see, the EstimationStrategy implies a TransformStrategy, so setting it should suffice, and from it the Config would set the TransformStrategy...

**Comment by glaurent** [comment]
@PCrompton To be honest, what you ask is a bit like asking how to play a Jazz solo having hardly ever touched an instrument. I only have vague recollection of basic signal processing classes from college, I could implement this algorithm only because I had two sample implementations that I could use as reference, and even then there was a whole lot of guesswork. Computing pitch information from a soundwave is actually still a research topic, and pretty hard to do, there's a lot of complex math involved. Get the basics of iOS development first, then you might want to tackle this kind of problems.

**Comment by vadymmarkov** [comment]
@PCrompton If you are familiar with iOS/Mac development basics, you can read about [Core audio](https://developer.apple.com/library/content/documentation/MusicAudio/Conceptual/CoreAudioOverview/CoreAudioEssentials/CoreAudioEssentials.html) and [AVFoundation](https://developer.apple.com/reference/avfoundation) to understand the concepts of working with audio. As for pitch detection, you can start with [Wikipedia](https://en.wikipedia.org/wiki/Pitch_detection_algorithm), there is a list of references that could possibly put you into the right direction. But it's a research topic, as @glaurent said, so normally I'm trying to discover the particular problem by reading different kinds of resources and/or exploring existing implementations in other programming languages.

**Comment by vadymmarkov** [comment]
@glaurent Yeah, I tend to use structs whenever it's not necessary to have a reference type 😄 I agree with you that the transform strategy is probably redundant in Config. At first I though it could be more flexible, but it seems like transformation and estimation go together anyway, so it might be something that I'd like to refactor.

**Comment by vadymmarkov** [comment]
I'm going to merge this PR. Will do some refactoring, as I mentioned before, and will make a new release straight after that. Great job @glaurent !

**Comment by glaurent** [comment]
@vadymmarkov  I tend to start from the opposite viewpoint (which AFAIK is more common) : you use classes unless it's necessary or convenient to have a value type. Anyway in this case it doesn't make much of a difference.

Thanks for the merge, you're welcome :). I'll let you close the pullreq.

**Pull request merged**